exago@1.4.1 +rocm build fails: tao/opflow_tao.cpp:220:10: error: no matching function for call to 'TaoGetObjective'

### Steps to reproduce the issue

`exago@1.4.1 +hiop +rocm amdgpu_target=gfx90a` build is failing using:
* `spack/develop` (4817e4c80ced0c1115a018bb185ce93f50b20242 from `Thu Apr 21 13:22:41 2022 -0700`)
* Ubuntu 20.04, x86_64
* GCC 11.1.0
* ROCm 4.5.2, installed outside of Spack and registered with Spack via external package definitions

It looks like maybe this is an issue of the `petsc` version used?

Spack environment: [spack.yaml.txt](https://github.com/spack/spack/files/8535500/spack.yaml.txt) (Contains ROCm 4.5.2 external definitions)

Concretization:
```
 -   jgvh36m  exago@1.4.1%gcc@11.1.0~cuda+hiop~ipo~ipopt+mpi+python~raja+rocm amdgpu_target=gfx90a build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
[+]  n2yk46c      ^cmake@3.23.1%gcc@11.1.0~doc+ncurses+ownlibs~qt build_type=Release arch=linux-ubuntu20.04-x86_64
[+]  3q2u23f          ^ncurses@6.2%gcc@11.1.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-x86_64
[+]  l32w2ng              ^pkgconf@1.8.0%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  o3xh65q          ^openssl@1.1.1n%gcc@11.1.0~docs~shared certs=system arch=linux-ubuntu20.04-x86_64
[+]  7tfv3dc              ^perl@5.34.1%gcc@11.1.0+cpanm+shared+threads arch=linux-ubuntu20.04-x86_64
[+]  n56hned                  ^berkeley-db@18.1.40%gcc@11.1.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu20.04-x86_64
[+]  er5k4wt                  ^bzip2@1.0.8%gcc@11.1.0~debug~pic+shared arch=linux-ubuntu20.04-x86_64
[+]  wo7wdk3                      ^diffutils@3.8%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  qnmvhfn                          ^libiconv@1.16%gcc@11.1.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  q73e6y3                  ^gdbm@1.19%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  q2tfszp                      ^readline@8.1%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  4fqlrsm                  ^zlib@1.2.12%gcc@11.1.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-x86_64
[+]  36yjkxt      ^hiop@0.6.0%gcc@11.1.0~cuda~deepchecking~ginkgo~ipo~jsrun~kron+mpi~raja+rocm~shared~sparse amdgpu_target=gfx90a build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
[+]  zqbz2yx          ^hip@4.5.2%gcc@11.1.0~ipo build_type=Release patches=24ca25a,2a41904 arch=linux-ubuntu20.04-x86_64
[+]  25wy5br          ^hipblas@4.5.2%gcc@11.1.0~ipo build_type=Release patches=a2cc151 arch=linux-ubuntu20.04-x86_64
[+]  h75nkkx          ^hipsparse@4.5.2%gcc@11.1.0~ipo build_type=Release arch=linux-ubuntu20.04-x86_64
[+]  qg2ri5r          ^hsa-rocr-dev@4.5.2%gcc@11.1.0+image~ipo+shared build_type=Release patches=71e6851 arch=linux-ubuntu20.04-x86_64
[+]  vmfznu5          ^llvm-amdgpu@4.5.2%gcc@11.1.0~ipo~link_llvm_dylib~llvm_dylib~openmp+rocm-device-libs build_type=Release patches=d999f3b arch=linux-ubuntu20.04-x86_64
[+]  eo3axh6          ^magma@2.6.2rc1%gcc@11.1.0~cuda+fortran~ipo+rocm+shared amdgpu_target=gfx1012 build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
[+]  wxywckr              ^openblas@0.3.20%gcc@11.1.0~bignuma~consistent_fpcsr~ilp64+locking+pic+shared symbol_suffix=none threads=openmp arch=linux-ubuntu20.04-x86_64
[+]  pgzwxng          ^mpich@4.0.2%gcc@11.1.0~argobots+fortran+hwloc+hydra+libxml2+pci+romio~slurm~two_level_namespace~verbs~wrapperrpath device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu20.04-x86_64
[+]  ajif2wp              ^findutils@4.8.0%gcc@11.1.0 patches=440b954 arch=linux-ubuntu20.04-x86_64
[+]  ezwimq7              ^hwloc@2.7.1%gcc@11.1.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared arch=linux-ubuntu20.04-x86_64
[+]  oqllgky                  ^libpciaccess@0.16%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  l26stqq                      ^libtool@2.4.7%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  v66vcke                          ^m4@1.4.19%gcc@11.1.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-x86_64
[+]  c7og75o                              ^libsigsegv@2.13%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  ov6i6jg                      ^util-macros@1.19.3%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  6srncvd                  ^libxml2@2.9.12%gcc@11.1.0~python arch=linux-ubuntu20.04-x86_64
[+]  hcbat4k                      ^xz@5.2.5%gcc@11.1.0+pic libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  dyjelsf              ^libfabric@1.14.0%gcc@11.1.0~debug~kdreg fabrics=rxm,sockets,tcp,udp arch=linux-ubuntu20.04-x86_64
[+]  4mv5yvx      ^petsc@3.17.0%gcc@11.1.0~X~batch~cgns~complex~cuda~debug+double~exodusii~fftw+fortran~giflib+hdf5~hpddm~hwloc+hypre~int64~jpeg~knl~kokkos~libpng~libyaml~memkind+metis~mkl-pardiso~mmg~moab~mpfr+mpi~mumps~openmp~p4est~parmmg~ptscotch~random123~rocm~saws~scalapack+shared~strumpack~suite-sparse+superlu-dist~tetgen~trilinos~valgrind clanguage=C arch=linux-ubuntu20.04-x86_64
[+]  dnayf5q          ^hdf5@1.12.1%gcc@11.1.0~cxx+fortran+hl~ipo~java+mpi+shared~szip~threadsafe+tools api=default build_type=RelWithDebInfo patches=ee351eb arch=linux-ubuntu20.04-x86_64
[+]  ropzdgl          ^hypre@2.24.0%gcc@11.1.0~complex~cuda~debug+fortran~gptune~int64~internal-superlu~mixedint+mpi~openmp~rocm+shared~superlu-dist~unified-memory arch=linux-ubuntu20.04-x86_64
[+]  mffh6rt          ^metis@5.1.0%gcc@11.1.0~gdb~int64~real64+shared build_type=Release patches=4991da9,b1225da arch=linux-ubuntu20.04-x86_64
[+]  ubkuie2          ^parmetis@4.0.3%gcc@11.1.0~gdb~int64~ipo+shared build_type=RelWithDebInfo patches=4f89253,50ed208,704b84f arch=linux-ubuntu20.04-x86_64
[+]  uthh4hm          ^python@3.8.12%gcc@11.1.0+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-ubuntu20.04-x86_64
[+]  jik6sl4              ^expat@2.4.8%gcc@11.1.0+libbsd arch=linux-ubuntu20.04-x86_64
[+]  vgxft26                  ^libbsd@0.11.5%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  mz4vout                      ^libmd@1.0.4%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  ues23dm              ^gettext@0.21%gcc@11.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-x86_64
[+]  klgel4y                  ^tar@1.34%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  t2pde6h              ^libffi@3.4.2%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  4xs4h3p              ^sqlite@3.37.2%gcc@11.1.0+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-ubuntu20.04-x86_64
[+]  vcg2y5m              ^util-linux-uuid@2.37.4%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  fywvfhb          ^superlu-dist@7.2.0%gcc@11.1.0~cuda~int64~ipo~openmp~rocm+shared build_type=RelWithDebInfo patches=8da9e21 arch=linux-ubuntu20.04-x86_64
[+]  ugouxwy      ^py-mpi4py@3.1.2%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  je4y5y4          ^py-pip@21.3.1%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  ibwaqga          ^py-setuptools@59.4.0%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
[+]  loa6wtx              ^py-wheel@0.37.0%gcc@11.1.0 arch=linux-ubuntu20.04-x86_64
```

Error is quickly reproduced here using cached dependency binaries and Docker container image `ecpe4s/ubuntu20.04-rocm:4.5.2`, which contains ROCm 4.5.2 components pre-installed such that they are ready-to-use with the Spack environment posted above.

```
# Copy the spack.yaml from above into the container first

$> git clone https://github.com/spack/spack
$> (cd spack && git checkout 4817e4c8)
$> . spack/share/spack/setup-env.sh

$> spack mirror add ExaGo https://cache.e4s.io/exago
$> spack buildcache keys -it

$> spack -e . concretize
...

$> spack -e . install --only dependencies --cache-only --include-build-deps
# Takes ~2 minutes on a fast connection

$> spack -e . install --no-cache
...
==> Installing exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e
==> No patches needed for exago
==> exago: Executing phase: 'cmake'
==> exago: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

5 errors found in build log:
     976     /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/include/petsctao.h:241:1: note: 'TaoSetHessianRoutine' has been explicitly marked
              deprecated here
     977     PETSC_DEPRECATED_FUNCTION("Use TaoSetHessian() (since version 3.17)") static inline PetscErrorCode TaoSetHessianRoutine(Tao t,Mat H,Mat P,PetscErrorCode (*f)(Tao,Vec,Mat,Mat,void*)
             ,void *c) { return TaoSetHessian(t,H,P,f,c); }
     978     ^
     979     /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/include/petscconf.h:11:53: note: expanded from macro 'PETSC_DEPRECATED_FUNCTION'
     980     #define PETSC_DEPRECATED_FUNCTION(why) __attribute((deprecated))
     981                                                         ^
  >> 982     /tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-src/src/opflow/solver/tao/opflow_tao.cpp:220:10: error: no matching function for call to 'TaoGe
             tObjective'
     983       ierr = TaoGetObjective(tao->nlp, obj);
     984              ^~~~~~~~~~~~~~~
     985     /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/include/petsctao.h:230:29: note: candidate function not viable: requires 3 argume
             nts, but 2 were provided
     986     PETSC_EXTERN PetscErrorCode TaoGetObjective(Tao,PetscErrorCode(**)(Tao,Vec,PetscReal*,void*),void**);
     987                                 ^
     988     /tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-src/src/opflow/interface/opflow.cpp:2769:5: warning: Use SETERRQ() (since version 3.17) [-W#pra
             gma-messages]

     ...

     995     #define PETSC_DEPRECATED_MACRO(why) _Pragma(why)
     996                                         ^
     997     <scratch space>:11:6: note: expanded from here
     998      GCC warning "Use SETERRQ() (since version 3.17)"
     999          ^
     1000    2 warnings generated when compiling for gfx90a.
  >> 1001    /tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-src/src/opflow/solver/tao/opflow_tao.cpp:230:10: error: use of undeclared identifier 'TaoGetSol
             utionVector'
     1002      ierr = TaoGetSolutionVector(tao->nlp, X);
     1003             ^
     1004    3 warnings and 2 errors generated when compiling for gfx90a.
     1005    3 warnings generated when compiling for gfx90a.
  >> 1006    make[2]: *** [src/opflow/CMakeFiles/OPFLOW_obj_static.dir/build.make:219: src/opflow/CMakeFiles/OPFLOW_obj_static.dir/solver/tao/opflow_tao.cpp.o] Error 1
     1007    make[2]: *** Waiting for unfinished jobs....
     1008    /tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-src/src/opflow/solver/hiop/opflow_hiop.cpp:424:5: warning: Use SETERRQ() (since version 3.17) [
             -W#pragma-messages]
     1009        SETERRQ1(PETSC_COMM_SELF, PETSC_ERR_SUP,
     1010        ^
     1011    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/include/petscerror.h:71:23: note: expanded from macro 'SETERRQ1'
     1012    #define SETERRQ1(...) PETSC_DEPRECATED_MACRO("GCC warning \"Use SETERRQ() (since version 3.17)\"") SETERRQ(__VA_ARGS__)

     ...

     1096    <scratch space>:55:6: note: expanded from here
     1097     GCC warning "Use SETERRQ() (since version 3.17)"
     1098         ^
     1099    3 warnings generated when compiling for host.
     1100    1 warning generated when compiling for host.
     1101    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-build-jgvh36m'
  >> 1102    make[1]: *** [CMakeFiles/Makefile2:779: src/opflow/CMakeFiles/OPFLOW_obj_static.dir/all] Error 2
     1103    make[1]: *** Waiting for unfinished jobs....
     1104    [ 45%] Linking CXX executable test_pflow_functionality
     1105    cd /tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-build-jgvh36m/tests/functionality/pflow && /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.
             1.0/cmake-3.23.1-n2yk46ck5ckqimr2f743in34htag3bxi/bin/cmake -E cmake_link_script CMakeFiles/test_pflow_functionality.dir/link.txt --verbose=1
     1106    /opt/rocm-4.5.2/hip/bin/hipcc -O2 -g -DNDEBUG -Wl,-rpath -Wl,/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/mpich-4.0.2-pgzwxngke6u7ecvnbm5jmihtqnyi27da/lib CMakeFiles/test_p
             flow_functionality.dir/selfcheck.cpp.o -o test_pflow_functionality  -Wl,-rpath,/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/li
             b:/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/mpich-4.0.2-pgzwxngke6u7ecvnbm5jmihtqnyi27da/lib:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
             ::::::::::::::::::::::::: ../../../src/pflow/libexago_pflow.a ../../../src/ps/libexago_ps.a ../../../src/utils/libexago_utils.a ../../../tpl/spdlog/libspdlog.a -lpthread /spack/opt
             /spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/lib/libpetsc.so /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/mpich-4.0.2-pgzwxngke6
             u7ecvnbm5jmihtqnyi27da/lib/libmpicxx.so /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/mpich-4.0.2-pgzwxngke6u7ecvnbm5jmihtqnyi27da/lib/libmpi.so
     1107    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-build-jgvh36m'
     1108    [ 45%] Built target test_pflow_functionality
     1109    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-build-jgvh36m'
  >> 1110    make: *** [Makefile:149: all] Error 2
```

 

### Error message

<details><summary>Error message</summary><pre>
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

5 errors found in build log:
     976     /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/include/petsctao.h:241:1: note: 'TaoSetHessianRoutine' has been explicitly marked
              deprecated here
     977     PETSC_DEPRECATED_FUNCTION("Use TaoSetHessian() (since version 3.17)") static inline PetscErrorCode TaoSetHessianRoutine(Tao t,Mat H,Mat P,PetscErrorCode (*f)(Tao,Vec,Mat,Mat,void*)
             ,void *c) { return TaoSetHessian(t,H,P,f,c); }
     978     ^
     979     /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/include/petscconf.h:11:53: note: expanded from macro 'PETSC_DEPRECATED_FUNCTION'
     980     #define PETSC_DEPRECATED_FUNCTION(why) __attribute((deprecated))
     981                                                         ^
  >> 982     /tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-src/src/opflow/solver/tao/opflow_tao.cpp:220:10: error: no matching function for call to 'TaoGe
             tObjective'
     983       ierr = TaoGetObjective(tao->nlp, obj);
     984              ^~~~~~~~~~~~~~~
     985     /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/include/petsctao.h:230:29: note: candidate function not viable: requires 3 argume
             nts, but 2 were provided
     986     PETSC_EXTERN PetscErrorCode TaoGetObjective(Tao,PetscErrorCode(**)(Tao,Vec,PetscReal*,void*),void**);
     987                                 ^
     988     /tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-src/src/opflow/interface/opflow.cpp:2769:5: warning: Use SETERRQ() (since version 3.17) [-W#pra
             gma-messages]

     ...

     995     #define PETSC_DEPRECATED_MACRO(why) _Pragma(why)
     996                                         ^
     997     <scratch space>:11:6: note: expanded from here
     998      GCC warning "Use SETERRQ() (since version 3.17)"
     999          ^
     1000    2 warnings generated when compiling for gfx90a.
  >> 1001    /tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-src/src/opflow/solver/tao/opflow_tao.cpp:230:10: error: use of undeclared identifier 'TaoGetSol
             utionVector'
     1002      ierr = TaoGetSolutionVector(tao->nlp, X);
     1003             ^
     1004    3 warnings and 2 errors generated when compiling for gfx90a.
     1005    3 warnings generated when compiling for gfx90a.
  >> 1006    make[2]: *** [src/opflow/CMakeFiles/OPFLOW_obj_static.dir/build.make:219: src/opflow/CMakeFiles/OPFLOW_obj_static.dir/solver/tao/opflow_tao.cpp.o] Error 1
     1007    make[2]: *** Waiting for unfinished jobs....
     1008    /tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-src/src/opflow/solver/hiop/opflow_hiop.cpp:424:5: warning: Use SETERRQ() (since version 3.17) [
             -W#pragma-messages]
     1009        SETERRQ1(PETSC_COMM_SELF, PETSC_ERR_SUP,
     1010        ^
     1011    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/include/petscerror.h:71:23: note: expanded from macro 'SETERRQ1'
     1012    #define SETERRQ1(...) PETSC_DEPRECATED_MACRO("GCC warning \"Use SETERRQ() (since version 3.17)\"") SETERRQ(__VA_ARGS__)

     ...

     1096    <scratch space>:55:6: note: expanded from here
     1097     GCC warning "Use SETERRQ() (since version 3.17)"
     1098         ^
     1099    3 warnings generated when compiling for host.
     1100    1 warning generated when compiling for host.
     1101    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-build-jgvh36m'
  >> 1102    make[1]: *** [CMakeFiles/Makefile2:779: src/opflow/CMakeFiles/OPFLOW_obj_static.dir/all] Error 2
     1103    make[1]: *** Waiting for unfinished jobs....
     1104    [ 45%] Linking CXX executable test_pflow_functionality
     1105    cd /tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-build-jgvh36m/tests/functionality/pflow && /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.
             1.0/cmake-3.23.1-n2yk46ck5ckqimr2f743in34htag3bxi/bin/cmake -E cmake_link_script CMakeFiles/test_pflow_functionality.dir/link.txt --verbose=1
     1106    /opt/rocm-4.5.2/hip/bin/hipcc -O2 -g -DNDEBUG -Wl,-rpath -Wl,/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/mpich-4.0.2-pgzwxngke6u7ecvnbm5jmihtqnyi27da/lib CMakeFiles/test_p
             flow_functionality.dir/selfcheck.cpp.o -o test_pflow_functionality  -Wl,-rpath,/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/li
             b:/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/mpich-4.0.2-pgzwxngke6u7ecvnbm5jmihtqnyi27da/lib:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
             ::::::::::::::::::::::::: ../../../src/pflow/libexago_pflow.a ../../../src/ps/libexago_ps.a ../../../src/utils/libexago_utils.a ../../../tpl/spdlog/libspdlog.a -lpthread /spack/opt
             /spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/petsc-3.17.0-4mv5yvxbje3xzmohqtg7uqkny4i4f3w6/lib/libpetsc.so /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/mpich-4.0.2-pgzwxngke6
             u7ecvnbm5jmihtqnyi27da/lib/libmpicxx.so /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/mpich-4.0.2-pgzwxngke6u7ecvnbm5jmihtqnyi27da/lib/libmpi.so
     1107    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-build-jgvh36m'
     1108    [ 45%] Built target test_pflow_functionality
     1109    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-exago-1.4.1-jgvh36mffeba6vo25ldki5w6gy6k5z6e/spack-build-jgvh36m'
  >> 1110    make: *** [Makefile:149: all] Error 2
</pre></details>


### Information on your system

* **Spack:** 0.18.0.dev0 (4817e4c80ced0c1115a018bb185ce93f50b20242)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/8535690/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/8535691/spack-build-out.txt)

@CameronRutherford @ashermancinelli

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/exago/package.py
+++ b/var/spack/repos/builtin/packages/exago/package.py
@@ -75,7 +75,7 @@ class Exago(CMakePackage, CudaPackage, ROCmPackage):
     depends_on('hiop+mpi', when='+hiop+mpi')
 
     depends_on('petsc@3.13:3.14', when='@:1.2.99')
-    depends_on('petsc@3.16.0:3.16.99', when='@1.3.0:')
+    depends_on('petsc@3.16:3.16', when='@1.3.0:')
     depends_on('petsc~mpi', when='~mpi')
 
     depends_on('ipopt', when='+ipopt')

--- a/var/spack/repos/builtin/packages/exago/package.py
+++ b/var/spack/repos/builtin/packages/exago/package.py
@@ -75,7 +75,7 @@ class Exago(CMakePackage, CudaPackage, ROCmPackage):
     depends_on('hiop+mpi', when='+hiop+mpi')
 
     depends_on('petsc@3.13:3.14', when='@:1.2.99')
-    depends_on('petsc@3.16.0:', when='@1.3.0:')
+    depends_on('petsc@3.16.0:3.16.99', when='@1.3.0:')
     depends_on('petsc~mpi', when='~mpi')
 
     depends_on('ipopt', when='+ipopt')

--- a/var/spack/repos/builtin/packages/exago/package.py
+++ b/var/spack/repos/builtin/packages/exago/package.py
@@ -75,7 +75,7 @@ class Exago(CMakePackage, CudaPackage, ROCmPackage):
     depends_on('hiop+mpi', when='+hiop+mpi')
 
     depends_on('petsc@3.13:3.14', when='@:1.2.99')
-    depends_on('petsc@3.16:3.16', when='@1.3.0:')
+    depends_on('petsc@3.16.0:3.16', when='@1.3.0:')
     depends_on('petsc~mpi', when='~mpi')
 
     depends_on('ipopt', when='+ipopt')

